src/QsCompiler/Transformations/Monomorphization.cs

```diff
@@ -87,13 +225,91 @@ public override TypedExpression Transform(TypedExpression ex)
             return base.Transform(ex);
         }
 
-        private string CreateTypePrefix(ImmutableDictionary<QsTypeParameter, ResolvedType> typeDict)
+        private Concretion CreateConcretion(ImmutableDictionary<QsTypeParameter, ResolvedType> typeDict)
         {
-            return "_" + String.Join("", typeDict
-                .ToList()
-                .OrderBy(kvp => kvp.Key.TypeName)
-                .Select(kvp => kvp.Key.TypeName.Value + "_" + kvp.Value.Resolution.ToString() + "_")
-                );
+            Concretion result = new Concretion();
+            foreach (var kvp in typeDict)
+            {
+                result.Add((kvp.Key, kvp.Value));
+            }
+            return result;
+        }
+
+        private Identifier GetConcreteIdentifier(Identifier.GlobalCallable globalCallable, ImmutableDictionary<QsTypeParameter, ResolvedType> types)
+        {
+            Concretion target = null;
+            if (types != null && !types.IsEmpty)
+            {
+                target = CreateConcretion(types);
+            }
+
+            string name = null;
+
+            // Check for recursive call
+            if (CurrentResponse.originalName.Namespace.Value == globalCallable.Item.Namespace.Value &&
+                CurrentResponse.originalName.Name.Value == globalCallable.Item.Name.Value &&
+                CurrentResponse.typeResolutions.SetEquals(target))
+            {
+                name = CurrentResponse.concreteCallable.FullName.Name.Value;
+            }
+
+            // Search Requests for identifier
+            if (name == null)
+            {
+                name = Requests
+                    .Where(req =>
+                        req.originalName.Namespace.Value == globalCallable.Item.Namespace.Value &&
```

Comment:
Since you repeatedly do similar comparisons I suggest to actually define a local function to do so.